Add initial implementation

Vinai opened the pull request:
The issue came up in the discussion of a PR to the Yireo_WebP compat module.

There currently is no way to request additional fields in GraphQL queries, that are usually embedded in .phtml templates.

Accessing additional related data is a common task in modules.

My first idea for a solution would be a service as a view model, that takes a GraphQL query string, parses it into a PHP array, dispatches an event and also provides a public method to plug into, and then returns the query as a string again.
In a template, instead of:

```php
$query = "
products(filter: {} pageSize: 20) {
  items {
    {$type}_products {
        sku
        id
        small_image {
          url
        }
    }
  }
}";
```

the code would look something like:

```php
$query = $gqlViewModel->query("product_list_query", "
products(filter: {} pageSize: 20) {
  items {
    {$type}_products {
        sku
        id
        small_image {
          url
        }
    }
  }
}");
```

The goal here would be to introduce as little complexity as possible from a developer perspective and keep backward compatibility.


A query customization is done with an event observer:

```php
$query = $event->getData('gql_container')->getData('query');

$gqlEditor = new GraphqlQueryEditor();

// add a single field to a result object
$query = $gqlEditor->setFieldIn($query, ['products', 'items', 'small_image'], 'url_webp');

// add multiple fields to a result object
$query = $gqlEditor->setFieldIn($query, ['products', 'items', 'image'], 'url label url_webp');

// add a query argument
$query = $gqlEditor->setArgumentIn($query, ['products', 'filter', 'name'], 'match', 'Tank');
$query = $gqlEditor->setArgumentIn($query, ['products'], 'pageSize', 2);

$event->getData('gql_container')->setData('query', $query);
```

Vinai committed:
Add initial implementation

Vinai commented on src/ViewModel/GraphqlViewModel.php:
```diff
@@ -0,0 +1,44 @@
+        return $this->toString($this->toAst($name, $query));
+    }
+
+    public function toAst(string $name, string $query): \GraphQL\Language\AST\DocumentNode
```
This method is public so it is pluginazable.
Maybe it would be better to have that on a different class, not the view model, because it is not intended to be called from templates or custom code.

schmengler replied:
Sounds reasonable, but is there really a need for plugins since we have the event? What's the use case?

Vinai replied:
Some people prefer plugins is the only use case I can think of...

schmengler replied:
I think we can be opinionated here, but it's not up to me to decide :)

schmengler commented on src/Model/GraphqlQueryEditor.php:
```diff
@@ -0,0 +1,274 @@
+     * Add or set argument to value at given path for GraphQL query or mutation.
+     *
+     * Examples:
+     * $sut->setArgumentIn($ast, ['products', 'filter', 'name'], 'match', 'Tank');
```
in the examples something like `$editor` would be more clear than `$sut`

Vinai replied:
Good catch, thanks!

schmengler commented on src/ViewModel/GraphqlViewModel.php:
```diff
@@ -0,0 +1,44 @@
+
+    private function toString(\GraphQL\Language\AST\DocumentNode $ast): string
+    {
+        return \GraphQL\Language\Printer::doPrint($ast);
```
@avstudnitz that might be interesting to you

Vinai committed:
Clarify example

Vinai committed:
Get rid of additional return

Vinai committed:
Make method private so it can not be used as a plugin target

Vinai committed:
Refactor to hide implementation detail

The benefits are:
- No overhead of parsing and reserializing of the query if
  no customizations are present.  This is my main reason for the change.
- Future changes of the parsing library can easily be integrated

The downsides of this change are:
- Customizing queries now requires getting the query string from a
  container object in the event, and setting the updated query on the
  container after the desired changes where made.
 - If multiple customizations need to be done the query is parsed and
   serialized once for each call to $editor->setFieldIn()
   and $editor->setArgumentIn().

Vinai commented:
I've refactored the PR so the internal parsed AST is no longer part of the API.  
It is now an implementation detail of the `GraphqlEditor`.

The benefits are:
- No overhead of parsing and reserializing of the query if no customizations are present.
  This is my main reason for the change.
- Future changes of the parsing library can easily be integrated

The downsides of this change are:
- Customizing queries now requires getting the query string from a container object in the event, and setting the updated query on the container after the desired changes where made.
-  If multiple customizations need to be done the query is parsed and serialized once for each call to `$editor->setFieldIn()` and `$editor->setArgumentIn()`.

I've updated the examples in the PR description accordingly.

Vinai committed:
Use query/mutation identifier as event suffix
